from snakemake.utils import validate
import pandas as pd
import numpy as np
import os
import datetime

sys.path.insert(1, 'scripts/')
sys.path.insert(2, '../workflow/scripts/')
import rnaseqpoptools as rnaseqpop

configfile: "config/config.yaml"
configpath = workflow.configfiles[0]

wkdir = os.getcwd()
metadata = rnaseqpop.load_metadata(config['metadata'])
validate(metadata, schema="schemas/samples.schema.yaml")
samples = metadata["sampleID"]

# AlleleBalance mutations file
if config["miscellaneous"]["VariantsOfInterest"]["activate"]:
    mutationData = pd.read_csv(
        config["miscellaneous"]["VariantsOfInterest"]["path"], sep="\t"
    )
else:
    mutationData = pd.DataFrame(
        {"Name": ["KDR"], "mut": ["995F"], "Location": ["2L-28xxxxx"]}
    )  # ignore this, just an dummy df for snakemake


include: "rules/common.smk"

welcome(version="v1.0.5")

include: "rules/qc.smk"
include: "rules/diffexp.smk"
include: "rules/variantsOfInterest.smk"
include: "rules/alignment.smk"
include: "rules/variantCalling.smk"
include: "rules/filterAnnotate.smk"
include: "rules/variantAnalysis.smk"
include: "rules/jupyter-book.smk"
include: "rules/misc.smk"


rule all:
    input:
        # The below function (defined in rules/common.smk) will read the config file and determine desired output files
        GetDesiredOutputs,
